(PUP-7961) Change packaging so that puppet-bolt does not conflict with an existing puppet

lutter opened the pull request:
We need to make sure that when this branch gets installed via a Ruby gem,
that we do not interfere with any other Puppet installations that the user
might already have. We do this by not mentioning 'lib/' in the require_path
of our gemspec. This effectively hides everything in 'lib/' from gem and
will keep ruby from ever satisfying a require of 'puppet/<somewhere>' with
files from this gem.

Code that consciously wants to load this gem, needs to instead require
'puppet_bolt'. That manipulates $LOAD_PATH so that 'lib/' does become
available. After that, any 'require "puppet/<somewhere>"' will resolve to a
file from this gem.

I have tried this out by making a gem with `rake package:gem` and installing the resulting gem in a container that did not have puppet. I then ran this script to verify that things seem on the up-and-up:

```ruby
begin
  require 'puppet'
  puts "Something is wrong: I can load puppet straightup"
rescue LoadError
  puts "No puppet in sight"
end

require 'puppet_bolt'
require 'puppet/version'

puts "After loading puppet_bolt, I have Puppet #{Puppet.version}"
```

puppetcla commented:
CLA signed by all contributors.

hlindberg commented:
A bit scary that the unit/integration tests (as run here by travis etc) does not test the change as it requires the gem to be built. So the green state is a different state. We also need to figure out how to do run from source like when it is packaged in a gem (for debugging etc). 
I have a suspicion that the change will break autoloader (3.x) and loaders (4x) when they load puppet language extensions from 'lib/puppet' (functions in Ruby).

MosesMendoza commented on ext/project_data.yaml:
```diff
@@ -1,5 +1,5 @@
 ---
-project: 'puppet'
+project: 'puppet-bolt'
```
i'm confused. why are we renaming the project to puppet-bolt?

Also @lutter could you open a ticket for this change? Though a small change-set it seems non-trivial enough to warrant one.

MosesMendoza commented:
Is this a temporary change that's only intended for this feature branch but will not ultimately be released in puppet?

lutter commented:
@hlindberg can you put a small example together that shows how this fails ? There shouldn't be any issue: once you 'require "puppet_bolt"', you are in the exact same state that gem usually puts you in when you 'require "puppet"' with the normal gem.

lutter commented:
@MosesMendoza yes, this is only meant for this branch. We want to be able to install a branch of Puppet as part of bolt that does not interfere with a 'normal' Puppet installation. Hence the different name. This should never be merged into master.

MosesMendoza commented:
Ah got it. Thanks

Iristyle commented:
Have you tried add a reference to the gem via `Gemfile` and doing a `bundle install`.  I would suggest trying with both a actual gem reference (temporarily publish it to a local Stickler) *and* via git style URI.

There are different gem load behaviors within Bundler.

lutter commented:
I just tried all that, by running the little example I mentioned above for each case (remembering to prefix `ruby /tmp/t.rb` with `bundle exec` ;)

I set up a local stickler put the puppet-bolt gem into it, and then mirrored everything from puppet's Gemfile.lock into it.

```bash
> gem install puppet-bolt
> ruby /tmp/t.rb
No puppet in sight
After loading puppet_bolt, I have Puppet 5.1.0-bolt
> gem uninstall puppet-bolt

> cat > Gemfile <<EOF
source 'http://localhost:6789/'

gem 'puppet-bolt'
EOF
> bundle install --path .bundle
> bundle exec ruby /tmp/t.rb
No puppet in sight
After loading puppet_bolt, I have Puppet 5.1.0-bolt

> rm -rf .bundle
> cat > Gemfile <<EOF
source 'http://localhost:6789/'

gem 'puppet-bolt', :git => "https://github.com/lutter/puppet.git", :branch => "pkg"
EOF
> bundle install --path .bundle
> bundle exec ruby /tmp/t.rb
No puppet in sight
After loading puppet_bolt, I have Puppet 5.1.0-bolt
```

The one change I had to make to this PR to make the last one work was to also update the `.gemspec`. That's also the reason why tests are failing now.

If we are ok with not being able to handle git style dependencies in a Gemfile, I would revert the changes to `.gemspec` so that the tests continue to work. People can always work around that with a local checkout and referring to that in their Gemfile.

haus commented:
@lutter what if someone has already loaded `puppet` from the standard load path and then they load `puppet_bolt`?

lutter commented:
@haus nobody besides the `bolt` command line executable should ever load `puppet_bolt` - the gem is not meant for general consumption, only for the purposes of `bolt`.

johnduarte commented on ext/project_data.yaml:
```diff
@@ -9,9 +9,10 @@ version_file: 'lib/puppet/version.rb'
-gem_files: '[A-Z]* install.rb bin lib conf man examples ext tasks spec locales'
+gem_files: '[A-Z]* install.rb bin lib libbolt conf man examples ext tasks spec locales'
 gem_test_files: 'spec/**/*'
 gem_executables: 'puppet'
```
This causes the gem to have executable conflicts with puppet when installed as a gem alongside puppet. I presume we should namespace or remove the executable.

johnduarte commented on ext/project_data.yaml:
```diff
@@ -9,9 +9,10 @@ version_file: 'lib/puppet/version.rb'
 gem_test_files: 'spec/**/*'
 gem_executables: 'puppet'
+gem_require_path: 'libbolt'
 gem_default_executables: 'puppet'
```
This causes the gem to have executable conflicts with puppet when installed as a gem alongside puppet. I presume we should namespace or remove the executable.

lutter commented:
@johnduarte I just updated the PR to no longer package `bin/` and removed any reference to the `puppet` executable. In my testing, this now seems to happily live side-by-side with an existing puppet installation.

Iristyle commented:
A bit of pedantic bookkeeping - do we have a PUP ticket to go along with this?  Otherwise when this branch later gets merged in to mainline, it could cause some trouble with the ticket reconciliation scripts.

Iristyle commented:
In documenting for you an explanation of how the `Gemfile` / `.gemspec` / `project_data.yaml` work in conjunction with one another, I've realized that it's no longer in the state that it should be (original changes to 4.10.1 back in April as part of https://github.com/puppetlabs/puppet/commit/19f78a20b70660b30fff8c0c49e8b9a9b3c99b40#diff-46ff332d66e3d0266138c1c6e3315453 are no longer present in `master`)

While I track down what has happened (bad merge perhaps?), I can at least clarify a bit on how the installed gemset was refactored to behave:

* For local source dev, bundler uses `Gemfile`, which follows the standard procedure of calling `gemspec` to load what's defined in the `.gemspec` for runtime dependencies (it has no explicit knowledge of `project_data.yaml`)
    * Under these circumstances, `project_data.yaml` is also loaded as the `.gemspec` loads it - this pulls in some project specific dependencies which cannot be statically defined in a `.gemspec`
* When specifying Puppet as a gem dependency in another project via `Gemfile` and a git style reference, `project_data.yaml` gets loaded via the same mechanism as a local Puppet source clone (this is how source based acceptance testing works - through `ci:test:git`)
* When building / packaging the gem, the rake tasks at https://github.com/puppetlabs/packaging/blob/master/tasks/gem.rake use a combo of what's statically defined in the `.gemspec` as a `runtime_dependency`, the rake tasks `.gemspec` template and `project_data.yaml`
    * This is necessary because we ship 4 flavors of the Puppet gem - 2 for Windows, 1 for darwin and the generic one.  Each one has a different set of dependencies, so that means there's no way to have a single `.gemspec` in the Puppet repo serve all the needs.  All dependencies have to be defined statically if you want them to appear correctly in places like rubygems for instance (there's a way to get dynamic dependencies in a gem, but it means you lose the visibility in rubygems, so it's not a good tradeoff)
* ALSO NOTE: There are some weird things that sometimes happen as a result of `BUNDLE_SPECIFIC_PLATFORM` *not* being passed into Bundler by default (more info in https://github.com/puppetlabs/puppet/pull/5780)

This likely won't materially impact this PR, but I think we still might want to get things back to the way they're supposed to be prior to merging this PR.  I'll keep you posted.

Iristyle commented:
PR is now up against 5.3.x branch to reinstate the PUP-7425 Gemfile / .gemspec refactorings - https://github.com/puppetlabs/puppet/pull/6206/files

That will need to be merged into master and merged up to this feature branch as well.

johnduarte commented:
@lutter the following line in the .gemspec appears to still cause the 'puppet' executable stub to get created

line 26 in your modified .gemspec file
```
s.executables = ["puppet"]
```

This triggers the executable conflict message when attempting to install puppet and puppet-bolt at the same time.

Reproduction steps on fresh debian-8-x86_64 system. All commands executed as root.
```
gpg --keyserver hkp://keys.gnupg.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 7D2BAF1CF37B13E2069D6956105BD0E739499BDB
\curl -sSL https://get.rvm.io | bash -s stable
rvm install 2.4
apt-get install git
git clone https://github.com/puppetlabs/puppet
cd puppet/
git checkout feature_puppet-tasks
curl -O https://patch-diff.githubusercontent.com/raw/puppetlabs/puppet/pull/6163.patch
git apply 6163.patch
gem build .gemspec
gem install puppet-bolt-5.2.0.gem
gem install puppet
```

lutter commented:
@Iristyle just created PUP-7911 for bookkeeping

@johnduarte the way things are set up, you will only get the right thing when you run `rake package:gem` but not when you run `gem build`. That's the two ways of building a puppet gem I mentioned the other day: I think it would be desirable to make 'gem build' still spit out a `puppet` gem so that the Travis/Appveyor tests still work, but have `rake package:gem` produce the `puppet-bolt` gem we want to ship.

I just updated the PR to reflect those things.

pcarlisle commented:
Do travis and appveyor really use the gem? I don't see it in the logs there. I'm running into problems using this as a local development dependency with bundler, because i want to require puppet-bolt in the gemspec and then override it in Gemfile, but since the puppet .gemspec still says the gem name is puppet bundler won't accept this.

Iristyle commented:
Travis / AppVeyor always do a git clone of the source - then a `bundle install`.  They don't build or use gems.  The scenario you're describing is my 2nd bullet in https://github.com/puppetlabs/puppet/pull/6163#issuecomment-329857811 @pcarlisle

Iristyle commented:
@lutter do you want to change `puppet` at https://github.com/puppetlabs/puppet/blob/master/.gemspec#L15 to `puppet-bolt`?  Also if you could rebase that would pick up some of the other dev workflow fixes that have since been merged back into this branch.

Thanks!

Iristyle commented:
Spiked some updates to this PR at https://github.com/puppetlabs/puppet/pull/6220

Seemed to work OK inside the `bolt` Gemfile locally.  Not sure where the problems were with the `.gemspec` modified as `puppet-bolt` @lutter?

lutter commented:
@Iristyle whenever I changed `.gemspec`, the tests in travis and appveyor blew up. We'll see if that doesn't happen with your change.

I also just realized that I screwed up the ticket number: it should be PUP-7961 (updated it here)

lutter commented:
@Iristyle I've now also updated this PR to incorporate your changes to `.gemspec` and rebased to the current HEAD of `feature-puppet_tasks`.

Iristyle commented:
We've encountered some issues when installing this new `puppet-bolt` gem into environments where the `puppet-agent` is installed.

We're rethinking the strategy of building / shipping a separate `puppet-bolt` gem and investigating vendoring the `puppet` code on this branch instead via a git submodule, and shipping that inside the `bolt` gem.  That removes the need for shipping an independent / short-lived `puppet-bolt` gem, but also involves adding to the bolt `.gemspec` to satisfy the other dependencies of `puppet`, `facter` and `hiera` - some of which are platform specific.

joshcooper commented:
This also breaks bolt when running from source, because it tries to install puppet from the feature branch, but puppet's gemspec doesn't satisfy the dependency (since it's name is different). I'm going to revert this, since it's causing bolt PRs to fail.
